doc/designs/lms.md

```diff
@@ -0,0 +1,243 @@
+  [2] requires that key and signature generation be performed in
+  hardware cryptographic modules that do not allow secret keying material to be
+  exported, even in encrypted form.
+- LMS and LM_OTS signatures.
+  The code will be written in a way that would allow LMS signatures to be added
+  if required, but a HSS signature with only a single level tree should be
+  sufficient for this use case. A separate LMS signature dispatch table could be
+  added if this is ever required.
+- CMS and X509 support - this may be added at a later date.
+- Implementing Extended Merkle signature scheme (XMSS) is a separate effort.
+
+HSS Signature and Key formats
+-----------------------------
+
+The LMS/HSS signature and public key formats use XDR (not ASN1 data).
+The reference code in RFC 8554 checks sizes of the binary data before loading
+the known fields. Rather than do this the OpenSSL PACKET interface is used.
+When the end of the data is expected PACKET_remaining() is tested.
+NOTE that the size of signatures depends on the modes which are embedded in the
+loaded data.
+Byte array data such as encoded public keys are just kept as pointers into the
+data. The entire signature blob is duplicated if the pointers are required
+across API calls.
+
+HSS message streaming API
+-------------------------
+
+In order to handle HSS message streaming the signature data must be decoded
+first, since the message is only used in the last part of the verify.
+This means that the existing sequence of calls for streaming (i.e.
+DigestVerifyInit/DigestVerifyUpdate/DigestFinal) is not sufficient for this
+purpose unless the message is buffered during the update, and then only
+processed during the final.
+
+So we should:
+
+- Support one shot operations:
+  Using existing EVP_DigestVerifyInit_ex + EVP_DigestVerify calls
+- Dont support streaming via existing interface:
+  EVP_DigestVerifyInit_ex + EVP_DigestVerifyUpdate + EVP_DigestVerifyFinal.
+  It could be done, but requires buffering of the message during updates.
+- EVP_VerifyInit etc should not be supported.
+- In order to set the signature early one of the following methods should be
+  used:
+
+Proposal 1:
+
+Add streaming via a new API that is similar to the existing DigestVerify API's,
+but moves the signature |sig| from the final to the init. i.e.
+
+```c
+__owur int EVP_DigestVerifyHBSInit(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
+                                   const char *mdname,
+                                   OSSL_LIB_CTX *libctx,
+                                   const char *props, EVP_PKEY *pkey,
+                                   const OSSL_PARAM params[],
+                                   const unsigned char *sig,
+                                   size_t siglen);   
+/* Alias: since the function signature is identical */
+#define EVP_DigestVerifyHBSUpdate(a,b,c) EVP_DigestVerifyUpdate(a,b,c)
+
+__owur int EVP_DigestVerifyHBSFinal(EVP_MD_CTX *ctx);
```

Comment:
I like this proposal more, except for:
1. It should not be called with HBS in the name as this is generic call for any possible future signature schemes and not tightly coupled to hash based signatures.
2. Why do we need both pkey as input parameter and ppctx as output one? Would it be better to have only one of them. I'm unsure which one of them though.
3. How does it play with this proposal https://github.com/openssl/openssl/blob/master/doc/designs/fetching-composite-algorithms.md

Reply:
1) I was trying to make the Init and FInal calls match... using a matching tag.. I dont care what the tag is.
2) It is the same API as the existing one with sig added to the end.
3) We would either need yet another function, or pass the signature in as an OSSL_PARAM.. We dont actually use EVP_SIGNATURE_fetch() normally currently, so this seems a bit early to be adding the EVP_SIGNATURE one.

Reply:
> I was trying to make the Init and FInal calls match... using a matching tag.. I dont care what the tag is.

Definitely needs a different name.

> We would either need yet another function, or pass the signature in as an OSSL_PARAM.. We dont actually use EVP_SIGNATURE_fetch() normally currently, so this seems a bit early to be adding the EVP_SIGNATURE one.

I don't see why?

Reply:
do_sigver_init() does this internally. Not much else needs the EVP_SIGNATURE_fetch()

Reply:
Actually, the 3rd option (passing `EVP_SIGNATURE`) will become available with the effort mentioned...  and it's now underway, so it wouldn't be a bad idea to collaborate around designing the API.  If we're reusing `do_sigver_init()` (not sure why we'd have to), it can be modified to fit.

Reply:
Err how can I comment on this design document?

Reply:
You can start a discussion or raise an issue...  or wait for the draft PR I'm currently preparing.

Do note that I'm currently exploring `EVP_PKEY_sign()` / `EVP_PKEY_verify()` and associated functions rather than `EVP_DigestSign` and `EVP_DigestVerify` function groups.

Reply:
When I think of it, I think I'll start with a design document, and we can discuss that.

Reply:
See #22672